plat/common/arm/traps_arm64.c

```diff
@@ -0,0 +1,210 @@
+DECLARE_TRAP_EVENT(UKARCH_TRAP_SECURITY);
+
+static const struct {
+	struct uk_event *event;
+	const char *str;
+} _trap_table[AARCH64_TRAP_MAX] = {
+	{ UK_EVENT_PTR(UKARCH_TRAP_INVALID_OP), "invalid op"     },
+	{ UK_EVENT_PTR(UKARCH_TRAP_DEBUG),      "debug"          },
+	{ UK_EVENT_PTR(UKARCH_TRAP_PAGE_FAULT), "page fault"     },
+	{ UK_EVENT_PTR(UKARCH_TRAP_BUS_ERROR),  "bus error"      },
+	{ UK_EVENT_PTR(UKARCH_TRAP_MATH),       "floating point" },
+	{ UK_EVENT_PTR(UKARCH_TRAP_SECURITY),   "security"       }
+};
+
+static enum aarch64_trap esr_to_trap(unsigned long esr)
+{
+	__u8 fsc, trap;
+
+	/* We expect Unikraft to run in EL1. So do not catch traps from EL0. */
+	switch (ESR_EC_FROM(esr)) {
+	case ESR_EL1_EC_MMU_IABRT_EL1:
+	case ESR_EL1_EC_MMU_DABRT_EL1:
+		fsc = ESR_ISS_ABRT_FSC_FROM(ESR_ISS_FROM(esr));
+		UK_ASSERT(fsc < NUM_ABORT_MAP_ENTRIES);
+
+		trap = _abort_map[fsc];
+		if (trap == 0)
+			break; /* Neither page fault, nor bus error */
+
+		return trap;
+
+	case ESR_EL1_EC_PC_ALIGN:
+	case ESR_EL1_EC_SP_ALIGN:
+		return AARCH64_TRAP_BUS_ERROR;
+
+	case ESR_EL1_EC_SVE_ASIMD_FP_ACC:
+	case ESR_EL1_EC_SVE_ACC:
+	case ESR_EL1_EC_FP64:
+		return AARCH64_TRAP_MATH;
+
+	case ESR_EL1_EC_UNKNOWN:
+	case ESR_EL1_EC_ILL:
+		return AARCH64_TRAP_INVALID_OP;
+
+	case ESR_EL1_EC_BTI:
+	case ESR_EL1_EC_FPAC:
+		return AARCH64_TRAP_SECURITY;
+
+	case ESR_EL1_EC_BRK_EL1:
+	case ESR_EL1_EC_STEP_EL1:
+	case ESR_EL1_EC_WATCHP_EL1:
+	case ESR_EL1_EC_BRK64:
+		return AARCH64_TRAP_DEBUG;
+	}
+
+	return AARCH64_TRAP_MAX;
+}
+
+static void dump_registers(struct __regs *regs, uint64_t far)
+{
+	unsigned char idx;
+
+	uk_pr_crit("Unikraft: Dump registers:\n");
+	uk_pr_crit("\t SP       : 0x%016lx\n", regs->sp);
+	uk_pr_crit("\t ESR_EL1  : 0x%016lx\n", regs->esr_el1);
+	uk_pr_crit("\t ELR_EL1  : 0x%016lx\n", regs->elr_el1);
+	uk_pr_crit("\t LR (x30) : 0x%016lx\n", regs->lr);
+	uk_pr_crit("\t PSTATE   : 0x%016lx\n", regs->spsr_el1);
+	uk_pr_crit("\t FAR_EL1  : 0x%016lx\n", far);
+
+	for (idx = 0; idx < 28; idx += 4)
+		uk_pr_crit("\t x%02d ~ x%02d: 0x%016lx 0x%016lx 0x%016lx 0x%016lx\n",
+			   idx, idx + 3, regs->x[idx], regs->x[idx + 1],
+			   regs->x[idx + 2], regs->x[idx + 3]);
+
+	uk_pr_crit("\t x28 ~ x29: 0x%016lx 0x%016lx\n",
+		   regs->x[28], regs->x[29]);
+}
+
+void invalid_trap_handler(struct __regs *regs, uint32_t el,
+				uint32_t reason, uint64_t far)
+{
+	uk_pr_crit("Unikraft: EL%d invalid %s trap caught\n",
+		   el, exception_modes[reason]);
+	dump_registers(regs, far);
+	ukplat_crash();
+}
+
+void trap_el1_sync(struct __regs *regs, uint64_t far)
+{
+	struct ukarch_trap_ctx ctx = {regs, regs->esr_el1, 1, 0, far};
+	enum aarch64_trap trap = esr_to_trap(regs->esr_el1);
+
+	if (trap < AARCH64_TRAP_MAX) {
+		if (uk_raise_event_ptr(_trap_table[trap].event, &ctx))
+			return;
+	}
+
+	uk_pr_crit("Unikraft: EL1 sync trap caught\n");
+
+	dump_registers(regs, far);
+	ukplat_crash();
+}
+
+void trap_el1_irq(void)
+{
+	if (!gic) {
```

Comment:
I replaced this with a simple `UK_ASSERT`. I think it can be considered a programming error if the GIC has not been initialized and the boot process was allowed to continue.